src/sync/PushManager/index.js

```diff
@@ -0,0 +1,201 @@
+  // variable used on browser to reconnect only when a new client was added, saving some authentication and sse connections.
+  let connectForNewClient = false;
+
+  /** PushManager functions related to initialization */
+
+  const reauthBackoff = new Backoff(connectPush, settings.authRetryBackoffBase);
+  const sseReconnectBackoff = new Backoff(sseClient.reopen.bind(sseClient), settings.streamingReconnectBackoffBase);
+
+  let timeoutId = 0;
+
+  function scheduleTokenRefresh(issuedAt, expirationTime) {
+    // clear scheduled token refresh if exists (needed when resuming PUSH)
+    if (timeoutId) clearTimeout(timeoutId);
+
+    // Set token refresh 10 minutes before expirationTime
+    const delayInSeconds = expirationTime - issuedAt - SECONDS_BEFORE_EXPIRATION;
+
+    timeoutId = setTimeout(connectPush, delayInSeconds * 1000);
+  }
+
+  function connectPush() {
+    log.info('Connecting to push streaming.');
+
+    const userKeys = clientContexts ? Object.keys(clientContexts) : undefined;
+    authenticate(settings, userKeys).then(
+      function (authData) {
+        // restart backoff retry counter for auth and SSE connections, due to HTTP/network errors
+        reauthBackoff.reset();
+        sseReconnectBackoff.reset(); // reset backoff in case SSE conexion has opened after a HTTP or network error.
+
+        // emit PUSH_DISCONNECT if org is not whitelisted
+        if (!authData.pushEnabled) {
+          log.error('Streaming is not enabled for the organization. Switching to polling mode.');
+          pushEmitter.emit(PUSH_DISCONNECT); // there is no need to close sseClient (it is not open on this scenario)
+          return;
+        }
+
+        // don't open SSE connection if a new shared client was added, since it means that a new authentication is taking place
+        if (userKeys && userKeys.length < Object.keys(clientContexts).length) {
+          return;
+        }
+
+        // Connect to SSE and schedule refresh token
+        const decodedToken = authData.decodedToken;
+        sseClient.open(authData);
+        scheduleTokenRefresh(decodedToken.iat, decodedToken.exp);
+      }
+    ).catch(
+      function (error) {
+
+        sseClient.close(); // no harm if already disconnected
+        pushEmitter.emit(PUSH_DISCONNECT); // no harm if `PUSH_DISCONNECT` was already notified
+
+        if (error.statusCode) {
+          switch (error.statusCode) {
+            case 401: // invalid api key
+              log.error(`Fail to authenticate for push notifications, with error message: "${error.message}".`);
+              return;
+          }
+        }
+
+        // Branch for other HTTP and network errors
+        const delayInMillis = reauthBackoff.scheduleCall();
+        log.error(`Fail to authenticate for push notifications, with error message: "${error.message}". Attempting to reauthenticate in ${delayInMillis / 1000} seconds.`);
+      }
+    );
+  }
+
+  function disconnectPush() {
+    sseClient.close();
+
```

Comment:
Log an info level statement, to mark the disconnection. Would use another one for marking the connection.

Disregard if you're already doing it somewhere else.

Reply:
Same, let's make sure (if you already did disregard but haven't seen a reaction to this comment indicating it)